Remove feature gated enum variants

tcharding opened the pull request:
This is the updated version of #874 (which I closed, force pushed, and then was unable to re-open - my bad).

Feature gating enum variants makes code that uses the library brittle while we do not have `non_exhaustive`, we should avoid doing so. Instead we can add a dummy type that is available when the feature is not turned on. Doing so enables the compiler to enforce that we do not create the error type that is feature gated when the feature is not enabled.

Remove the feature gating around `bitcoinconsensus` error enum variants.

Closes: #645

apoelstra commented:
Needs rebase because of #806, sorry

tcharding commented:
No changes in force push, rebase only.

dr-orlovsky reviewed (commented):
Regarding derives on dumb error types: while we should not _underderive_, we can overderive. I.e., if the upstream adds derives to the respective error types this will not be a breaking change for them, but it will break our code. Thus, I suggest to do all possible derives for dumb errors, namely `Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Debug` (`Default` for errors is a bad practice, I assume).

apoelstra reviewed (approved):
ACK f4d015e46653a481d41ed06b025e5d090172b1cc

Though I am also a little confused why we're dropping `derive`s in this PR.

tcharding commented:
Because of this comment: https://github.com/rust-bitcoin/rust-bitcoin/pull/874#discussion_r823613829

dr-orlovsky commented:
> Because of this comment: https://github.com/rust-bitcoin/rust-bitcoin/pull/874#discussion_r823613829

I do not agree with that. Commented there

tcharding commented:
Oh, I thought your comment was just in the general case. I'll put some more thought into digesting the whole thread of discussion again. Thanks for re-stating you position.

tcharding commented:
Rebased on master and added all derives as suggested by @dr-orlovsky (+ Clone).

> Though I am also a little confused why we're dropping derives in this PR.

PR no longer does this.

sanket1729 reviewed (approved):
tACK 6ad2902814bd37902932c40778a7ec108c62737e. This is an improvment. 

But I think the correct solution here is to have a separate error type different from `script::Error` for the `verify` and `verify_with_flags` methods. Maybe this should be dealt with #820
